File: ReactNative/HUMAN_Demo/ios/HUMAN_Demo/HumanModule.m
#import "HumanModule.h"
@import HUMAN;

static NSString *humanNewHeaders = @"HumanNewHeaders";
static NSString *humanChallengeResult = @"HumanChallengeResult";
static NSString *humanSolved = @"solved";
static NSString *humanCancelled = @"cancelled";
static NSString *humanFalse = @"false";

@implementation HumanModule

// MARK: - NSObject

- (instancetype)init {
  self = [super init];
  shared = self;
  return self;
}

// MARK: - RCTEventEmitter

- (NSArray<NSString *> *)supportedEvents {
  return @[humanNewHeaders, humanChallengeResult];
}

// MARK: - PX Module

static HumanModule *shared = nil;

+ (HumanModule *)shared {
  return shared;
}

+ (BOOL)requiresMainQueueSetup {
  return YES;
}

- (void)handleUpdatedHeaders:(NSDictionary<NSString *,NSString *> *)headers {
  NSData *data = [NSJSONSerialization dataWithJSONObject:headers options:0 error:nil];
  NSString *json = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
  [self sendEventWithName:humanNewHeaders body:json];
}

- (void)handleChallengeSolvedEvent {
  [self sendEventWithName:humanChallengeResult body:humanSolved];
}

- (void)handleChallengeCancelledEvent {
  [self sendEventWithName:humanChallengeResult body:humanCancelled];
}

RCT_EXPORT_MODULE(HumanModule);

RCT_REMAP_METHOD(getHTTPHeaders,
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject) {
  NSDictionary<NSString *, NSString *> *headers = [HumanSecurity.BD headersForURLRequestForAppId:nil];
  NSData *data = [NSJSONSerialization dataWithJSONObject:headers options:0 error:nil];
  NSString *json = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
  resolve(@[json]);
}

RCT_REMAP_METHOD(handleResponse,
                 rsponse:(NSString *)response
                 code:(NSInteger)code
                 url:(NSString *)url
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject) {
  NSData *data = [response dataUsingEncoding:NSUTF8StringEncoding];
  NSHTTPURLResponse *httpURLResponse = [[NSHTTPURLResponse alloc] initWithURL:[NSURL URLWithString:url] statusCode:code HTTPVersion:nil headerFields:nil];
  BOOL handled = [HumanSecurity.BD handleResponseWithResponse:httpURLResponse data:data callback:^(enum HSBotDefenderChallengeResult result) {
    resolve((result == HSBotDefenderChallengeResultSolved ? humanSolved : humanCancelled));
  }];
  if (!handled) {
    resolve(humanFalse);
  }
}

@end

File: ReactNative/HUMAN_Demo/android/app/src/main/java/com/human_demo/HumanModule.java
package com.human_demo;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.humansecurity.mobile_sdk.HumanSecurity;
import com.humansecurity.mobile_sdk.main.HSBotDefenderChallengeResult;

import org.json.JSONObject;

import java.util.HashMap;

public class HumanModule extends ReactContextBaseJavaModule {

    static HumanModule shared = null;

    String humanNewHeaders = "HumanNewHeaders";
    String humanChallengeResult = "HumanChallengeResult";
    String humanSolved = "solved";
    String humanCancelled = "cancelled";
    String humanFalse = "false";

    // PX Module

    HumanModule(ReactApplicationContext context) {
        super(context);
    }

    @NonNull
    @Override
    public String getName() {
        return "HumanModule";
    }

    public void handleUpdatedHeaders(HashMap<String, String> headers) {
        if (!this.getReactApplicationContext().hasCatalystInstance()) {
            return;
        }
        JSONObject json = new JSONObject(headers);
        this.getReactApplicationContext()
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(humanNewHeaders, json.toString());
    }

    public void handleChallengeSolvedEvent() {
        if (!this.getReactApplicationContext().hasCatalystInstance()) {
            return;
        }
        this.getReactApplicationContext()
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(humanChallengeResult, humanSolved);
    }

    public void handleChallengeCancelledEvent() {
        if (!this.getReactApplicationContext().hasCatalystInstance()) {
            return;
        }
        this.getReactApplicationContext()
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(humanChallengeResult, humanCancelled);
    }

    @ReactMethod
    public void getHTTPHeaders(Promise promise) {
        JSONObject json = new JSONObject(HumanSecurity.INSTANCE.getBD().headersForURLRequest(null));
        promise.resolve(json.toString());
    }

    @ReactMethod
    public void handleResponse(String response, Integer code, String url, Promise promise) {
        boolean handled = HumanSecurity.INSTANCE.getBD().handleBlockResponse(response, url, result -> {
            promise.resolve(result == HSBotDefenderChallengeResult.SOLVED ? humanSolved : humanCancelled);
            return null;
        });
        if (!handled) {
            promise.resolve(humanFalse);
        }
    }
}
